cni-plugin/deployment/scripts/install-cni.sh

```diff
@@ -163,58 +202,148 @@ contexts:
+  local cni_conf_path="$1"
+ 
+  create_cni_conf
+  local tmp_data=""
+  local conf_data=""
+  if [ -e "${cni_conf_path}" ]; then
+   # Add the linkerd-cni plugin to the existing list
+   tmp_data=$(cat "${TMP_CONF}")
+   conf_data=$(jq --argjson CNI_TMP_CONF_DATA "${tmp_data}" -f /linkerd/filter.jq "${cni_conf_path}")
+   echo "${conf_data}" > ${TMP_CONF}
+  fi
+
+  # If the old config filename ends with .conf, rename it to .conflist, because it has changed to be a list
+  filename=${cni_conf_path##*/}
+  extension="${filename##*.}"
+  if [ "${filename}" != '01-linkerd-cni.conf' ] && [ "${extension}" = 'conf' ]; then
+   echo "Renaming ${cni_conf_path} extension to .conflist"
+   cni_conf_path="${cni_conf_path}list"
+  fi
+
+  if [ -e "${DEFAULT_CNI_CONF_PATH}" ] && [ "$cni_conf_path" != "${DEFAULT_CNI_CONF_PATH}" ]; then
+   echo "Removing Linkerd's configuration file: ${DEFAULT_CNI_CONF_PATH}"
+   rm -f "${DEFAULT_CNI_CONF_PATH}"
+  fi
+
+  # Move the temporary CNI config into place.
+  mv "${TMP_CONF}" "${cni_conf_path}" || exit_with_error 'Failed to mv files.'
+
+  echo "Created CNI config ${cni_conf_path}"
+}
+
+# Sync() is responsible for reacting to file system changes. It is used in
+# conjunction with inotify events; sync() is called with the name of the file that
+# has changed, the event type (which can be either 'CREATE' or 'DELETE'), and
+# the previously observed SHA of the configuration file.
+#
+# Based on the changed file and event type, sync() might re-install the CNI
+# plugin's configuration file.
+sync() {
+  local filename="$1"
+  local ev="$2"
+  local filepath="${HOST_CNI_NET}/$filename"
+
+  local prev_sha="$3"
+
+  local config_file_count
+  local new_sha
+  if [ "$ev" = "DELETE" ]; then
+    # When the event type is 'DELETE', we check to see if there are any `*conf` or `*conflist`
+    # files on the host's filesystem. If none are present, we install in
+    # 'interface' mode, using our own CNI config file.
+    config_file_count=$(find "${HOST_CNI_NET}" -maxdepth 1 -type f \( -iname '*conflist' -o -iname '*conf' \) | sort | wc -l)
+    if [ "$config_file_count" -eq 0 ]; then
+      echo "No active CNI configuration file found after $ev event; re-installing in \"interface\" mode"
+      install_cni_conf "${DEFAULT_CNI_CONF_PATH}"
+    fi
+  elif [ "$ev" = "CREATE" ]; then
+    # When the event type is 'CREATE', we check the previously observed SHA (updated
+    # with each file watch) and compare it against the new file's SHA. If they
+    # differ, it means something has changed.
+    new_sha=$(sha256sum "${filepath}" | awk '{print $1}')
+    if [ "$new_sha" != "$prev_sha" ]; then
+      # Create but don't rm old one since we don't know if this will be configured
+      # to run as _the_ cni plugin.
+      echo "New file [$filename] detected; re-installing in \"chained\" mode"
+      install_cni_conf "$filepath"
+    fi
+  else
+    # If the SHA hasn't changed or we get an unrecognised event, ignore it.
```

Comment:
technically, I don't think we hit this case for CREATE events where the sha doesn't change.  Do you think we should add some logging in an else block to the she comparison, above?